android/autodispose-android-archcomponents/src/androidTest/java/com/uber/autodispose/android/lifecycle/KotlinExtensionsTest.kt

```diff
@@ -0,0 +1,323 @@
+    assertThat(o.takeNext()).isEqualTo(0)
+
+    subject.onNext(1)
+    assertThat(o.takeNext()).isEqualTo(1)
+
+    lifecycle.emit(Lifecycle.Event.ON_PAUSE)
+    lifecycle.emit(Lifecycle.Event.ON_STOP)
+
+    subject.onNext(2)
+    assertThat(o.takeNext()).isEqualTo(2)
+
+    // We should stop here
+    lifecycle.emit(Lifecycle.Event.ON_DESTROY)
+    subject.onNext(3)
+    o.assertNoMoreEvents()
+  }
+
+  @Test
+  @UiThreadTest
+  fun observable_offAfterPause_shouldStopOnStop() {
+    val o = RecordingObserver<Int>(LOGGER)
+    val subject = PublishSubject.create<Int>()
+
+    val lifecycle = TestLifecycleOwner.create()
+    lifecycle.emit(Lifecycle.Event.ON_CREATE)
+    lifecycle.emit(Lifecycle.Event.ON_START)
+    lifecycle.emit(Lifecycle.Event.ON_RESUME)
+    lifecycle.emit(Lifecycle.Event.ON_PAUSE)
+    subject.autoDisposable(lifecycle).subscribe(o)
+
+    val d = o.takeSubscribe()
+
+    subject.onNext(2)
+    assertThat(o.takeNext()).isEqualTo(2)
+
+    // We could resume again
+    lifecycle.emit(Lifecycle.Event.ON_RESUME)
+    subject.onNext(3)
+    assertThat(o.takeNext()).isEqualTo(3)
+
+    // We should stop here
+    lifecycle.emit(Lifecycle.Event.ON_STOP)
+    subject.onNext(3)
+    o.assertNoMoreEvents()
+  }
+
+  @Test
+  @UiThreadTest
+  fun observable_offAfterOnDestroyView_shouldStopOnDestroy() {
+    val o = RecordingObserver<Int>(LOGGER)
+    val subject = PublishSubject.create<Int>()
+
+    val lifecycle = TestLifecycleOwner.create()
+    lifecycle.emit(Lifecycle.Event.ON_CREATE)
+    lifecycle.emit(Lifecycle.Event.ON_START)
+    lifecycle.emit(Lifecycle.Event.ON_RESUME)
+    lifecycle.emit(Lifecycle.Event.ON_PAUSE)
+    lifecycle.emit(Lifecycle.Event.ON_STOP)
+    // In a CREATED state now but the next event will be destroy
+    // This simulates subscribing in fragments' onDestroyView, where we want the subscription to
+    // still dispose properly in onDestroy.
+    subject.autoDisposable(lifecycle).subscribe(o)
+
+    val d = o.takeSubscribe()
+
+    subject.onNext(2)
+    assertThat(o.takeNext()).isEqualTo(2)
+
+    // We should stop here
+    lifecycle.emit(Lifecycle.Event.ON_DESTROY)
+    subject.onNext(3)
+    o.assertNoMoreEvents()
+  }
+
+  @Test
+  fun observable_offMainThread_shouldFail() {
+    val o = RecordingObserver<Int>(LOGGER)
+    val subject = PublishSubject.create<Int>()
+
+    // Spin it up
+    val lifecycle = TestLifecycleOwner.create()
+    lifecycle.emit(Lifecycle.Event.ON_CREATE)
+    lifecycle.emit(Lifecycle.Event.ON_START)
+    lifecycle.emit(Lifecycle.Event.ON_RESUME)
+    subject.autoDisposable(lifecycle).subscribe(o)
+
+    val d = o.takeSubscribe()
+    val t = o.takeError()
+    assertThat(t).isInstanceOf(IllegalStateException::class.java)
+    assertThat(t.message).contains("main thread")
+    o.assertNoMoreEvents()
+    assertThat(d.isDisposed).isTrue()
+  }
+
+  @Test(expected = OutsideScopeException::class)
+  @UiThreadTest
+  fun observable_offAfterDestroy_shouldFail() {
```

Comment:
This test is the only one where the behavior is different. 
Since we're trying to get the corresponding `Lifecycle.Event` to the `Lifecycle.State` super early (while trying to resolve the default `untilEvent`), during `ON_DESTROY` the `OutsideScopeException` will be called immediately.

Reply:
hmm, can we defer the resolution of `untilEvent` through something like `UntilEventFunction` and `CorrespondingEventsFunction` in the existing java code? I'd be fine making that 

Hadn't seen this before approving, let's try to keep the behavior the same before merging

Reply:
That's the part that is tricky (and probably not possible). `untilEvent` has to be resolved immediately since it has a default value. In our case it becomes a conundrum because we need it to be deferred.